pkg/controllers/deprovisioning/types.go

```diff
@@ -0,0 +1,158 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+)
+
+// Result is used to indicate the action of consolidating so we can optimize by not trying to consolidate if
+// we were unable to consolidate the cluster and it hasn't changed state with respect to pods/nodes.
+type Result byte
+
+const (
+	ResultNothingToDo Result = iota // there are no actions that can be performed given the current cluster state
+	ResultRetry                     // we attempted an action, but its validation failed so retry soon
+	ResultFailed                    // the action failed entirely
+	ResultSuccess                   // the action was successful
+)
+
+func (r Result) String() string {
+	switch r {
+	case ResultNothingToDo:
+		return "Nothing to do"
+	case ResultRetry:
+		return "Retry"
+	case ResultFailed:
+		return "Failed"
+	case ResultSuccess:
+		return "Success"
+	default:
+		return fmt.Sprintf("Unknown (%d)", r)
+	}
+}
+
+type deprovisioner interface {
+	// shouldNotBeDeprovisioned is a predicate used to filter deprovisionable nodes
+	shouldNotBeDeprovisioned(context.Context, *state.Node, *v1alpha5.Provisioner, []*v1.Pod) bool
+	// sortCandidates orders deprovisionable nodes by the deprovisioner's pre-determined priority
+	sortCandidates([]candidateNode) []candidateNode
+	// computeCommand generates a deprovisioning command given deprovisionable nodes
+	computeCommand(context.Context, int, ...candidateNode) (deprovisioningCommand, error)
+	// validateCommand validates a command for a deprovisioner
+	validateCommand(context.Context, []candidateNode, deprovisioningCommand) (bool, error)
+	// isExecutableCommand checks that a command can be executed by the deprovisioner
+	isExecutableCommand(deprovisioningCommand) bool
+	// getTTL returns the time to wait for a deprovisioner's validation
+	getTTL() time.Duration
+	// string is the string representation of the deprovisioner
+	string() string
+}
+
+type action byte
+
+const (
+	actionUnknown action = iota
+	actionNotPossible
+	actionDeleteConsolidation
+	actionReplaceConsolidation
+	actionDeleteEmpty
+	actionDoNothing
+	actionFailed
+)
+
+func (a action) getMetricsReasonName() string {
+	if a == actionDeleteConsolidation || a == actionReplaceConsolidation {
+		return metrics.ConsolidationReason
+	} else if a == actionDeleteEmpty {
+		return metrics.EmptinessReason
+	}
+	return ""
+}
+
+func (a action) needsReplacement() bool {
+	return a == actionReplaceConsolidation
+}
+
+func (a action) String() string {
```

Comment:
I agree, was hoping to keep this structure the same since I didn't want to mess with metrics that much, but as it looks like I'm changing them anyway, I'll make the change.